cpp/src/community/leiden_kernels.cu

```diff
@@ -0,0 +1,289 @@
+  //
+  rmm::device_vector<weight_t> vertex_weights_v(graph.number_of_vertices);
+  rmm::device_vector<weight_t> cluster_weights_v(graph.number_of_vertices);
+  rmm::device_vector<vertex_t> cluster_v(graph.number_of_vertices);
+
+  //
+  //  Temporaries used within kernels.  Each iteration uses less
+  //  of this memory
+  //
+  rmm::device_vector<vertex_t> tmp_arr_v(graph.number_of_vertices);
+  rmm::device_vector<vertex_t> cluster_inverse_v(graph.number_of_vertices);
+
+  weight_t total_edge_weight =
+    thrust::reduce(rmm::exec_policy(stream)->on(stream), weights_v.begin(), weights_v.end());
+  weight_t best_modularity = -1;
+
+  //
+  //  Initialize every cluster to reference each vertex to itself
+  //
+  thrust::sequence(rmm::exec_policy(stream)->on(stream), cluster_v.begin(), cluster_v.end());
+  thrust::copy(cluster_v.begin(), cluster_v.end(), cluster_vec);
+
+  //
+  //  Our copy of the graph.  Each iteration of the outer loop will
+  //  shrink this copy of the graph.
+  //
+  GraphCSRView<vertex_t, edge_t, weight_t> current_graph(offsets_v.data().get(),
+                                                         indices_v.data().get(),
+                                                         weights_v.data().get(),
+                                                         graph.number_of_vertices,
+                                                         graph.number_of_edges);
+
+  current_graph.get_source_indices(src_indices_v.data().get());
+
+  while (num_level < max_level) {
+    //
+    //  Sum the weights of all edges departing a vertex.  This is
+    //  loop invariant, so we'll compute it here.
+    //
+    //  Cluster weights are equivalent to vertex weights with this initial
+    //  graph
+    //
+#ifdef TIMING
+    hr_timer.start("init");
+#endif
+
+    cugraph::detail::compute_vertex_sums(current_graph, vertex_weights_v, stream);
+    thrust::copy(vertex_weights_v.begin(), vertex_weights_v.end(), cluster_weights_v.begin());
+
+#ifdef TIMING
+    hr_timer.stop();
+
+    hr_timer.start("update_clustering");
+#endif
+
+    weight_t new_Q = update_clustering_by_delta_modularity(total_edge_weight,
+                                                           resolution,
+                                                           current_graph,
+                                                           src_indices_v,
+                                                           vertex_weights_v,
+                                                           cluster_weights_v,
+                                                           cluster_v,
+                                                           stream);
+
+    // After finding the initial unconstrained partition we use that partitioning as the constraint
+    // for the second round.
+    rmm::device_vector<vertex_t> constraint(graph.number_of_vertices);
+    thrust::copy(cluster_v.begin(), cluster_v.end(), constraint.begin());
```

Comment:
Consider using `thrust::copy(rmm::exec_policy(stream)->on(stream),...)` that carries stream information.